conu/backend/nspawn/container.py

```diff
@@ -0,0 +1,261 @@
+
+import functools
+import logging
+import subprocess
+import time
+import glob
+
+
+from conu.apidefs.container import Container
+from conu.apidefs.filesystem import Filesystem
+from conu.exceptions import ConuException
+from conu.utils import check_port, run_cmd, random_str, output_decode
+from conu.utils.probes import Probe
+import constants
+
+
+logger = logging.getLogger(__name__)
+
+
+class NspawnContainer(Container):
+    systemd_wait_support = None
+
+    def __init__(self, image, container_id, name=None, popen_instance=None):
+        super(NspawnContainer, self).__init__(image, container_id, name)
+        if not name:
+            self.name = self._id
+        self.popen_instance = popen_instance
+
+    def __repr__(self):
+        return "NspawnContainer(image=%s, id=%s)" % (self.image, self.get_id())
+
+    def __str__(self):
+        return self.get_id()
+
+    def get_id(self):
+        return self._id
+
+    def inspect(self, refresh=True):
+        """
+        return cached metadata by default (a convenience method)
+
+        :param refresh: bool, returns up to date metadata if set to True
+        :return: dict
+        """
+        return self.get_metadata(refresh=refresh)
+
+    def get_metadata(self, refresh=True):
+        """
+        return cached metadata by default
+
+        :param refresh: bool, returns up to date metadata if set to True
+        :return: dict
+        """
+        if refresh or not self._metadata:
+            ident = self._id or self.name
+            if not ident:
+                raise ConuException("This container does not have a valid identifier.")
+            self._metadata = output_decode(run_cmd(["machinectl", "show",  ident], return_output=True))
+        return self._metadata
+
+    def is_running(self):
+        return "running" == self.get_metadata(refresh=True)["State"]
+
+    def copy_to(self, src, dest):
+        """
+        copy a file or a directory from host system to a container
+
+        :param src: str, path to a file or a directory on host system
+        :param dest: str, path to a file or a directory within container
+        :return: None
+        """
+        # using `docker cp` b/c put_archive is too complicated
+        logger.debug("copying %s from host to container at %s", src, dest)
+        cmd = ["machinectl", "copy-to",self.get_id(), src, dest]
+        run_cmd(cmd)
+
+    def copy_from(self, src, dest):
+        """
+        copy a file or a directory from container or image to host system.
+
+        :param src: str, path to a file or a directory within container or image
+        :param dest: str, path to a file or a directory on host system
+        :return: None
+        """
+        # using `docker cp` b/c get_archive is too complicated
+        logger.debug("copying %s from host to container at %s", src, dest)
+        cmd = ["machinectl", "copy-from", self.get_id(), src, dest]
+        run_cmd(cmd)
+
+
+    def stop(self):
```

Comment:
Should stop(), kill() and delete() point to a single method?  Could we take inspiration from `DockerContainer.stop()`, call poweroff to machine, wait for timeout and then call kill()? Would it be possible?